libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -776,10 +818,54 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
 {
 	setLocation(_binOp);
 
-	solAssert(!!_binOp.annotation().commonType);
+	if (FunctionDefinition const* function = _binOp.annotation().userDefinedFunction)
+	{
+		_binOp.leftExpression().accept(*this);
+		_binOp.rightExpression().accept(*this);
+		setLocation(_binOp);
+
+		solAssert(
+			function->isFree() || function->libraryFunction(),
+			"Only file-level functions and library functions can be bound to a user type operator."
+		);
+
+		FunctionType const* functionType = dynamic_cast<FunctionType const*>(
+			function->libraryFunction() ? function->typeViaContractName() : function->type()
+		);
+		solAssert(functionType);
+		functionType = dynamic_cast<FunctionType const&>(*functionType).asBoundFunction();
+		solAssert(functionType);
+		solAssert(
+			functionType->parameterTypes().size() == 1,
+			"Only functions with one parameter other than self parameter can be bound to a user type binary operator."
+		);
+
+		string left = expressionAsType(_binOp.leftExpression(), *functionType->selfType());
+		string right = expressionAsType(_binOp.rightExpression(), *functionType->parameterTypes().at(0));
+		solAssert(!left.empty() && !right.empty());
+
+		solAssert(function->isImplemented(), "");
+
+		solAssert(
+			function->returnParameters().size() == 1,
+			"A function bound to the user type operator is supposed to return exactly one value."
+		);
+		solAssert(*_binOp.annotation().type == *function->returnParameters().at(0)->type(),
+			"A return type of the bound function is supposed to be same as a operator type."
+		);
+
+		define(_binOp) <<
+			m_context.enqueueFunctionForCodeGeneration(*function) <<
+			("(" + left + ", " + right + ")\n");
+
+		return false;
+	}
+
+	solAssert(!!_binOp.annotation().commonType, "");
```

Comment:
You know this already, because you did it already almost everywhere. It is not needed anymore to define the second empty parameter in `solAssert`. Maybe you can remove those already in all the lines that you changed in this PR.